### DISCUSSION TITLE
Remove Unused Imports on save

### DISCUSSION THREAD

**Discussion by larsschieffer** [LazyVim/LazyVim#1943] [Q&A] [answered]
Hi there 👋 

I would like to run the code action “Remove Unused Imports” automatically on save for my typescript projects.

The action originated from the LazyExtra [Typescript](https://www.lazyvim.org/extras/lang/typescript#nvim-lspconfig)

Thanks.

**Comment by yushaku** [comment]
i think you just need to add keybinding for save and remove unused import at same key:
```
{
  "<C-s>",
  function()
    vim.lsp.buf.code_action({
      apply = true,
      context = {
        only = { "source.removeUnused.ts" },
        diagnostics = {},
      },
    })
  end,
  desc = "Remove unused imports",
})
map({ "i", "v", "n" }, "<C-s>", "<cmd>w<cr><esc>", { desc = "Save file" })
```

**Reply by larsschieffer** [reply]
Thank you @yushaku  for your fast reply!

**Comment by dimas-cyriaco** [comment, marked as answer]
Maybe you could add a autocmd like this (didn't test):

```lua
vim.api.nvim_create_autocmd({ "BufWritePre" }, {
  group = vim.api.nvim_create_augroup("ts_imports", { clear = true }),
  pattern = { "typescript" },
  callback = function()
    vim.lsp.buf.code_action({
      apply = true,
      context = {
        only = { "source.removeUnused.ts" },
        diagnostics = {},
      },
    })
  end,
})
```

**Reply by larsschieffer** [reply]
Thank you, @dimas-cyriaco ! That's exactly what I'm looking for.

In the end, I extended the pattern for react files too and it works great.
```lua
-- Remove unused imports on save
vim.api.nvim_create_autocmd({ "BufWritePre" }, {
  group = vim.api.nvim_create_augroup("ts_imports", { clear = true }),
  pattern = { "*.tsx,*.ts" },
  callback = function()
    vim.lsp.buf.code_action({
      apply = true,
      context = {
        only = { "source.removeUnused.ts" },
        diagnostics = {},
      },
    })
  end,
})

**Reply by zbyju** [reply]
Am I the only one that has a problem with this that it leads to some unexpected behavior? It feels like it's fighting with the formatter and there are some race conditions or something and leads to some code getting deleted or weirdly formatted. Am I doing something wrong?
